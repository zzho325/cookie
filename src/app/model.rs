pub mod editor;
pub mod focus;
pub mod messages;
pub mod session;
pub mod session_manager;
pub mod setting_manager;

use crate::{
    app::model::{
        focus::{Focusable, Focused},
        session::Session,
        session_manager::SessionManager,
        setting_manager::SettingManager,
    },
    models::configs::Configs,
};

pub struct Model {
    pub configs: Configs,
    /// Current session.
    pub session: Session,
    pub session_manager: SessionManager,
    /// Source of truth for selected session id.
    /// It might not be in session_summaries for newly created session.
    /// It might not be the same as id in session while waiting for fetching selected session.
    pub selected_session_id: Option<uuid::Uuid>,

    pub setting_manager_popup: Option<SettingManager>,
    pub show_sidebar: bool,
    pub focused: Focused,
    focus_order: Vec<fn(&mut Model) -> &mut dyn Focusable>,
    pub should_quit: bool,
}

impl Model {
    pub fn new(configs: Configs) -> Self {
        // FIXME: fix config usage
        let default_llm_settings = configs.derive_llm_settings();

        let mut this = Self {
            configs,
            session: Session::new(default_llm_settings),
            session_manager: SessionManager::default(),
            selected_session_id: None,
            setting_manager_popup: None,
            show_sidebar: false,
            should_quit: false,
            focused: Focused::Session,
            focus_order: Vec::new(),
        };

        this.focus_order.push(|m| &mut m.session);
        this.focus_order.push(|m| &mut m.session_manager);
        this.focus_order[0](&mut this).set_focus(true);
        this
    }

    pub fn quit(&mut self) {
        self.should_quit = true;
    }

    pub fn toggle_sidebar(&mut self) {
        if !self.show_sidebar {
            self.shift_focus_to(Focused::SessionManager);
        } else {
            self.shift_focus_to(Focused::Session);
        }
        self.show_sidebar = !self.show_sidebar;
    }

    /// Opens an new empty chat and enables editing.
    pub fn new_draft_chat(&mut self) {
        self.session.reset(self.configs.derive_llm_settings());
        self.selected_session_id = None;
        self.session_manager.set_selected(None);
        self.shift_focus_to(Focused::Session);
        self.session.is_editing = true;
    }
}

#[cfg(test)]
mod tests {
    use uuid::Uuid;

    use crate::{
        app::model::{Model, focus::Focused, messages::Messages},
        models::{ChatEvent, ChatMessage, Message, Role, SessionSummary},
    };

    #[test]
    fn new_draft_chat() {
        // ----------------------------------------------------------------
        // Setup model.
        // ----------------------------------------------------------------
        let llm_settings = crate::models::settings::LlmSettings::OpenAI {
            model: crate::service::llms::open_ai::api::OpenAIModel::Gpt4o,
            web_search: false,
        };
        let session_id = Uuid::new_v4();
        let title = "Awesome chat".to_string();

        let mut messages = Messages::default();
        messages.handle_user_chat_message(ChatMessage::new(
            session_id,
            llm_settings.clone(),
            Role::User,
            "history question".to_string(),
        ));
        messages.handle_chat_event_stream(ChatEvent::new(
            session_id,
            llm_settings.clone(),
            Message {
                role: Role::Assistant,
                msg: "history reponse".to_string(),
            }
            .into(),
        ));
        messages.handle_user_chat_message(ChatMessage::new(
            session_id,
            llm_settings.clone(),
            Role::User,
            "pending question".to_string(),
        ));
        messages.scroll_down();

        let mut model = Model::new(crate::models::configs::Configs::default());
        model.session.set_title(Some(title.clone()));
        model.session.set_messages(messages);
        model.session.is_editing = false;
        *model.session.input_editor.input_mut() = "repeat this".repeat(3);
        model.selected_session_id = Some(session_id);

        model.session_manager.handle_session_summaries(
            vec![SessionSummary {
                id: session_id,
                title,
                updated_at: chrono::Utc::now(),
            }],
            model.selected_session_id,
        );

        // ----------------------------------------------------------------
        // Verify new chat behavior.
        // ----------------------------------------------------------------
        model.new_draft_chat();

        // Session history, pending messages and messages scroll srate are reset.
        assert_eq!(
            model.session.messages.chat_messages().len(),
            0,
            "history messages are reset"
        );
        assert!(
            !model.session.messages.is_pending(),
            "pending messages are reset"
        );
        assert_eq!(
            model.session.messages.scroll_state().scroll_offset(),
            (0u16, 0u16),
            "messages scroll state is reset"
        );

        // Session manager reset.
        assert!(
            model.selected_session_id.is_none(),
            "model selected session id is reset"
        );
        assert!(
            model.session_manager.list_state_mut().selected().is_none(),
            "session manager selected is reset"
        );

        // Input editor remains.
        assert!(
            !model.session.input_editor.input().is_empty(),
            "input remains"
        );

        // Focus on editor and in editing mode.
        assert_eq!(
            model.focused,
            Focused::Session,
            "focus is shifted to session"
        );
        assert!(model.session.is_editing, "editing is enabled")
    }
}
