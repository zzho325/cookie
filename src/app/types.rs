use crossterm::event::KeyCode;

/// drives update
pub enum Message {
    Key(KeyCode),
    ServiceResp(String),
    CrosstermClose,
}

/// side effect of update
pub enum Command {
    ServiceReq(String),
}
