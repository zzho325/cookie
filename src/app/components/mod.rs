pub mod chat;
mod messages;
