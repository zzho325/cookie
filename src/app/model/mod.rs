pub mod editor;
pub mod messages;
pub mod scroll;

use std::sync::Arc;

use crossterm::event::KeyEvent;

use crate::{
    app::model::{
        editor::{Editor, WrapMode},
        messages::Messages,
    },
    models::{LlmSettings, ServiceReq, ServiceResp},
};

pub struct Model {
    pub llm_settings: Arc<LlmSettings>,
    pub should_quit: bool,
    pub messages: Messages,
    pub input_editor: Editor,
}

impl Model {
    pub fn new(default_llm_settings: LlmSettings) -> Self {
        let shared_llm_settings = Arc::new(default_llm_settings);
        Self {
            llm_settings: shared_llm_settings.clone(),
            should_quit: false,
            messages: Messages::new(shared_llm_settings),
            // by default editting
            input_editor: Editor::new(String::new(), true, WrapMode::default()),
        }
    }
}

/// Drives update.
pub enum Message {
    Key(KeyEvent),
    ServiceResp(ServiceResp),
    Send,
    CrosstermClose,
}

/// Side effect of update.
pub enum Command {
    SendMessage(String),
}

impl Command {
    /// If this `Command` corresponds to a service request, return `Some(_)`, otherwise return `None`.
    pub fn into_service_req(self) -> Option<ServiceReq> {
        let Command::SendMessage(msg) = self;
        Some(ServiceReq::ChatMessage(msg))
    }
}
