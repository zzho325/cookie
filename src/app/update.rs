mod session;

use crossterm::event::{KeyCode, KeyEvent};

use crate::{
    app::{
        Command, Message,
        model::{Focused, Model},
    },
    models::ServiceResp,
};

pub type Update = (Option<Message>, Option<Command>);

/// Updates model with message and optionally creates next message for chained update and command
/// for side effect.
pub fn update(model: &mut Model, msg: Message) -> Update {
    match msg {
        Message::Key(evt) => return handle_key_event(model, evt),
        Message::ServiceResp(resp) => {
            return handle_service_resp(model, resp);
        }
        Message::Send => {
            if let Some(req) = model.session.handle_user_message() {
                return (None, Some(Command::ServiceReq(req)));
            }
        }
        Message::GetSession(id) => {}
        Message::CrosstermClose => {
            model.should_quit = true;
        }
    }
    (None, None)
}

fn handle_service_resp(model: &mut Model, resp: ServiceResp) -> Update {
    match resp {
        ServiceResp::ChatMessage(assistant_message) => {
            model.session.handle_assistant_message(assistant_message)
        }

        ServiceResp::Sessions(session_summaries) => model
            .session_manager
            .handle_sessions_update(session_summaries, model.session.session_id()),
        _ => todo!(),
    }
    (None, None)
}

fn handle_key_event(model: &mut Model, keyevent: KeyEvent) -> Update {
    match model.focused {
        Focused::Session => {
            return session::handle_session_key_event(model, keyevent);
        }
        Focused::SessionManager => match keyevent.code {
            KeyCode::Char('q') => model.quit(),
            KeyCode::Char('s') => model.toggle_sidebar(),
            KeyCode::Char('n') => {
                model.session.reset(model.configs.derive_llm_settings());
                model.shift_focus_to(Focused::Session);
            }
            KeyCode::Down | KeyCode::Char('j') => {
                model.session_manager.select_next();
                let maybe_msg = model
                    .session_manager
                    .selected()
                    .map(|id| Message::GetSession(id));
                return (maybe_msg, None);
            }
            KeyCode::Up | KeyCode::Char('k') => {
                model.session_manager.select_previous();
                let maybe_msg = model
                    .session_manager
                    .selected()
                    .map(|id| Message::GetSession(id));
                return (maybe_msg, None);
            }
            KeyCode::Tab => model.shift_focus(),
            _ => {}
        },
    }
    (None, None)
}

#[cfg(test)]
mod tests {
    use crossterm::event::{KeyCode, KeyEvent};

    use crate::{
        app::{
            model::{Focused, Model},
            update::handle_key_event,
        },
        models::configs::Configs,
    };

    #[test]
    fn navigation() {
        struct Case {
            description: &'static str,
            focused: Focused,
            show_sidebar: bool,
            is_editing: bool,
            key_event: KeyEvent,
            expected_focused: Focused,
            expected_show_sidebar: bool,
        }

        let cases = vec![
            Case {
                description: "key tab does not navigate focus when sidebar is hidden",
                focused: Focused::Session,
                show_sidebar: false,
                is_editing: false,
                key_event: KeyCode::Tab.into(),
                expected_focused: Focused::Session,
                expected_show_sidebar: false,
            },
            Case {
                description: "key s opens and navigates to sidebar",
                focused: Focused::Session,
                show_sidebar: false,
                is_editing: false,
                key_event: KeyCode::Char('s').into(),
                expected_focused: Focused::SessionManager,
                expected_show_sidebar: true,
            },
            Case {
                description: "key s closes and navigates away from sidebar",
                focused: Focused::SessionManager,
                show_sidebar: true,
                is_editing: false,
                key_event: KeyCode::Char('s').into(),
                expected_focused: Focused::Session,
                expected_show_sidebar: false,
            },
            Case {
                description: "Tab navigates from session manager to session",
                focused: Focused::SessionManager,
                show_sidebar: true,
                is_editing: false,
                key_event: KeyCode::Tab.into(),
                expected_focused: Focused::Session,
                expected_show_sidebar: true,
            },
        ];

        for case in cases {
            let mut model = Model::new(Configs::default());
            model.session.is_editing = case.is_editing;
            model.shift_focus_to(case.focused);
            model.show_sidebar = case.show_sidebar;
            handle_key_event(&mut model, case.key_event);
            assert_eq!(
                model.focused, case.expected_focused,
                "{} focused",
                case.description
            );
            assert_eq!(
                model.show_sidebar, case.expected_show_sidebar,
                "{} show_sidebar",
                case.description
            );
        }
    }
}
