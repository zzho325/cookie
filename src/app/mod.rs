pub mod types;

use color_eyre::{
    Result,
    eyre::{Context as _, bail},
};
use crossterm::event::{Event, EventStream, KeyCode, KeyEventKind};
use futures_util::stream::StreamExt;
use ratatui::{
    Frame,
    style::Stylize as _,
    symbols::border,
    text::{Line, Text},
    widgets::{Block, Paragraph},
};
use tokio::{
    select,
    sync::mpsc::{UnboundedReceiver, UnboundedSender},
};

use crate::app::types::{Command, Message};

#[derive(Debug, Default)]
pub struct Model {
    should_quit: bool,

    question: String,
    answer: String,
    is_editing: bool,
}

pub struct App {
    // frontend <> backend channels
    req_tx: UnboundedSender<String>,
    resp_rx: UnboundedReceiver<String>,
}

impl App {
    pub fn new(
        req_tx: UnboundedSender<String>,
        resp_rx: UnboundedReceiver<String>,
    ) -> Result<Self> {
        Ok(Self { req_tx, resp_rx })
    }

    /// runs the application's main loop until the user quits
    pub async fn run(&mut self) -> Result<()> {
        let mut terminal = ratatui::init();
        let mut model = Model::default();

        let mut event_reader = EventStream::new();

        while !model.should_quit {
            // draw first so we see the latest state immediately
            terminal.draw(|f| Self::view(&mut model, f))?;
            let mut maybe_msg = select! {
                // key event from Crossterm
                maybe_evt = event_reader.next() => {
                    Self::handle_crossterm_event(&model, maybe_evt)?
                }
                // service response
                maybe_resp = self.resp_rx.recv() => {
                    if let Some(resp) = maybe_resp {
                    Some(Message::ServiceResp(resp))
                    } else {
                        bail!("service stream closed");
                    }
                }
            };

            // handle chained messages and side effect from update
            while let Some(msg) = maybe_msg {
                let (next_msg, maybe_cmd) = Self::update(&mut model, msg);
                maybe_msg = next_msg;
                if let Some(Command::ServiceReq(req)) = maybe_cmd {
                    self.req_tx.send(req)?
                }
            }
        }

        Ok(())
    }

    /// convert crossterm event to message
    pub fn handle_crossterm_event(
        _: &Model,
        maybe_evt: Option<Result<Event, std::io::Error>>,
    ) -> Result<Option<Message>> {
        match maybe_evt {
            Some(Ok(Event::Key(key))) if key.kind == KeyEventKind::Press => {
                Ok(Some(Message::Key(key.code)))
            }
            Some(Ok(_)) => Ok(None),
            Some(Err(e)) => Err(e).context("reading crossterm event failed"),
            None => {
                // the EventStream has closed
                Ok(Some(Message::CrosstermClose))
            }
        }
    }

    /// update model with message and optionally create next message for chained update and command
    /// for side effect
    fn update(model: &mut Model, msg: Message) -> (Option<Message>, Option<Command>) {
        match msg {
            Message::Key(code) => Self::handle_key_code(model, code),
            Message::ServiceResp(msg) => {
                model.answer = msg.clone();
                (None, None)
            }
            Message::CrosstermClose => {
                model.should_quit = true;
                (None, None)
            }
        }
    }

    fn handle_key_code(model: &mut Model, code: KeyCode) -> (Option<Message>, Option<Command>) {
        if model.is_editing {
            match code {
                KeyCode::Esc => {
                    model.is_editing = false;
                }
                KeyCode::Enter => {
                    model.is_editing = false;
                    return (None, Some(Command::ServiceReq(model.question.clone())));
                }
                KeyCode::Char(c) => model.question.push(c),
                KeyCode::Backspace => {
                    model.question.pop();
                }
                _ => {}
            }
        } else {
            match code {
                KeyCode::Char('q') => model.should_quit = true,
                KeyCode::Char('i') => model.is_editing = true,
                _ => {}
            }
        }
        (None, None)
    }

    fn view(model: &mut Model, frame: &mut Frame) {
        let title = Line::from(" Cookie ".bold());
        let block = Block::bordered()
            .title(title.centered())
            .border_set(border::THICK);

        let texts = Text::from(vec![
            Line::from(vec!["Q: ".into(), model.question.clone().into()]),
            Line::from(vec!["A: ".into(), model.answer.clone().into()]),
        ]);

        frame.render_widget(Paragraph::new(texts).centered().block(block), frame.area());
    }
}
