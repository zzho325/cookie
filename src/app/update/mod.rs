mod session;

use crossterm::event::{KeyCode, KeyEvent};

use crate::{
    app::{
        Command, Message,
        model::{Focused, Model},
    },
    models::ServiceResp,
};

pub type Update = (Option<Message>, Option<Command>);

/// Updates model with message and optionally create next message for chained update and command
/// for side effect.
pub fn update(model: &mut Model, msg: Message) -> Update {
    match msg {
        Message::Key(evt) => return handle_key_event(model, evt),
        Message::ServiceResp(resp) => {
            return handle_service_resp(model, resp);
        }
        Message::Send => {
            if let Some(req) = model.session.handle_user_message() {
                return (None, Some(Command::ServiceReq(req)));
            }
        }
        Message::CrosstermClose => {
            model.should_quit = true;
        }
    }
    (None, None)
}

fn handle_service_resp(model: &mut Model, resp: ServiceResp) -> Update {
    match resp {
        ServiceResp::ChatMessage(assistant_message) => {
            model.session.handle_assistant_message(assistant_message)
        }

        ServiceResp::Sessions(session_summaries) => model
            .session_manager
            .handle_sessions_update(session_summaries),
        _ => todo!(),
    }
    (None, None)
}

fn handle_key_event(model: &mut Model, keyevent: KeyEvent) -> Update {
    match model.focused {
        Focused::Session => {
            return session::handle_session_key_event(model, keyevent);
        }
        Focused::SessionManager => match keyevent.code {
            KeyCode::Char('q') => model.quit(),
            KeyCode::Char('s') => model.toggle_sidebar(),
            KeyCode::Tab => model.shift_focus(),
            _ => {}
        },
    }
    (None, None)
}

#[cfg(test)]
mod tests {
    use crossterm::event::KeyCode;

    use crate::app::{
        model::{Focused, Model},
        update::handle_key_event,
    };

    #[test]
    fn navigation() {
        let mut model = Model::new(crate::models::LlmSettings::default());

        handle_key_event(&mut model, KeyCode::Tab.into());
        assert_eq!(
            model.focused,
            Focused::Session,
            "key tab does not navigate focus when sidebar is hidden"
        );

        handle_key_event(&mut model, KeyCode::Esc.into());
        assert!(!model.session.is_editing, "key Esc toggles editing");

        handle_key_event(&mut model, KeyCode::Char('s').into());
        assert!(model.show_sidebar, "key s toggles sidebar");
        assert_eq!(
            model.focused,
            Focused::SessionManager,
            "key s navigates to sidebar"
        );

        handle_key_event(&mut model, KeyCode::Tab.into());
        assert_eq!(model.focused, Focused::Session, "tab nagivates to session");
    }
}
