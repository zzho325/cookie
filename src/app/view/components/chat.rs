use ratatui::{
    buffer::Buffer,
    layout::{Constraint, Direction, Layout, Rect},
    style::{Color, Modifier, Style},
    text::{Line, Span, Text},
    widgets::{Block, Borders, Padding, Paragraph, StatefulWidget, Widget},
};

use crate::{
    app::{
        model::{editor::Editor, messages::Messages, scroll::Scrollable},
        view::{
            components::messages::MessagesView,
            constants::{
                BORDER_THICKNESS, BORDER_THICKNESS_SIDE, MAX_INPUT_RATIO, MIN_INPUT_HEIGHT,
            },
        },
    },
    models::LlmSettings,
};

pub struct ChatState {
    pub cursor_position: Option<(u16, u16)>,
}

pub struct ChatView<'a> {
    pub messages: &'a Messages,
    pub input_editor: &'a mut Editor,
    pub llm: &'a LlmSettings,
}

impl StatefulWidget for ChatView<'_> {
    type State = ChatState;

    /// Renders chat pane with input block starting with MIN_INPUT_HEIGHT including border and
    /// increase height as input length increases with a maximum of MAX_INPUT_RATIO of widget area.
    fn render(self, area: Rect, buf: &mut Buffer, state: &mut ChatState) {
        let input_content_width = area.width.saturating_sub(BORDER_THICKNESS as u16) as usize;
        self.input_editor.set_width(input_content_width);
        let max_input_height = (area.height as f32 * MAX_INPUT_RATIO).floor() as usize;
        let max_input_content_height = max_input_height - BORDER_THICKNESS_SIDE;
        self.input_editor.set_max_height(max_input_content_height);

        let lines = self.input_editor.lines();
        let input_height =
            (lines.len() + BORDER_THICKNESS_SIDE).clamp(MIN_INPUT_HEIGHT, max_input_height) as u16;
        let message_height = area.height.saturating_sub(input_height);

        let layout = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Length(message_height),
                Constraint::Length(input_height),
            ])
            .split(area);

        let messages = MessagesView::from(self.messages);
        messages.render(layout[0], buf);

        // input
        let input_lines: Vec<Line> = lines.into_iter().map(Line::from).collect();
        let text = Text::from(input_lines);

        // construct title
        // TODO: centralize style here and prompt style
        let provider = self.llm.provider_name();
        let model = self.llm.model_name();
        let title: Line = Line::from(vec![
            Span::raw("â”€ "),
            Span::styled(
                provider,
                Style::default()
                    .fg(Color::LightGreen)
                    .add_modifier(Modifier::BOLD),
            ),
            Span::raw(" on "),
            Span::styled(model, Style::default().fg(Color::LightBlue)),
            Span::raw(" "),
        ]);

        let scroll_offset = self.input_editor.scroll_offset();
        Paragraph::new(text)
            .block(
                Block::new()
                    .borders(Borders::TOP)
                    .padding(Padding::horizontal(1))
                    .title(title.left_aligned()),
            )
            .scroll(scroll_offset)
            .render(layout[1], buf);

        // set cursor position if editing
        if self.input_editor.is_editing {
            let cursor_position = self.input_editor.cursor_position();
            let (x, y) = cursor_position;
            let (y_scroll_offset, _) = scroll_offset;
            state.cursor_position = Some((
                x + BORDER_THICKNESS_SIDE as u16,
                y + message_height + BORDER_THICKNESS_SIDE as u16 - y_scroll_offset,
            ));
        } else {
            state.cursor_position = None;
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::app::{
        model::{
            editor::{Editor, WrapMode},
            messages::{HistoryMessage, MessageMetadata, Messages},
        },
        view::components::chat::ChatState,
    };

    #[test]
    fn render_chat() {
        let req_time = std::time::Instant::now();
        let resp_time = req_time + std::time::Duration::from_secs(2);

        let llm = crate::models::LlmSettings::OpenAI {
            model: crate::service::client::api::OpenAIModel::Gpt4o,
            web_search: false,
        };
        let history_llm = llm.clone();
        let history_message = HistoryMessage {
            user_msg: "history question".to_string(),
            assistant_msg: "history answer".to_string(),
            metadata: MessageMetadata {
                llm: history_llm,
                req_time,
                resp_time: Some(resp_time),
            },
        };

        let chat = super::ChatView {
            messages: &Messages {
                history_messages: vec![history_message],
                ..Messages::default()
            },
            input_editor: &mut Editor::new("repeat this".repeat(3), false, WrapMode::default()),
            llm: &llm,
        };
        let chat_state = &mut ChatState {
            cursor_position: None,
        };

        let mut terminal =
            ratatui::Terminal::new(ratatui::backend::TestBackend::new(20, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_stateful_widget(chat, frame.area(), chat_state))
            .unwrap();
        insta::assert_snapshot!(terminal.backend());
    }
}
