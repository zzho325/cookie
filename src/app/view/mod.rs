mod components;
pub mod constants;

use crate::app::{
    model::Model,
    view::components::chat::{ChatState, ChatView},
};
use ratatui::{Frame, layout::Position};

pub fn render_ui(model: &mut Model, frame: &mut Frame) {
    tracing::debug!("render_ui");
    let chat_state = &mut ChatState {
        cursor_position: None,
    };
    frame.render_stateful_widget(
        ChatView {
            is_editing: model.session.is_editing,
            messages: &model.session.messages,
            input_editor: &mut model.session.input_editor,
            llm_settings: &model.session.llm_settings,
        },
        frame.area(),
        chat_state,
    );

    if let Some((x, y)) = chat_state.cursor_position {
        frame.set_cursor_position(Position::new(x, y));
    }
}
