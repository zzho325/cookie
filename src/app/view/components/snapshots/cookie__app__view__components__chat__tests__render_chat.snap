---
source: src/app/view/components/chat.rs
expression: terminal.backend()
---
"                    "
"┌─> openAI on 4o    "
"[2s]                "
"└─> history question"
"history answer      "
"                    "
"                    "
"                    "
"                    "
"                    "
"                    "
"                    "
"                    "
"                    "
"                    "
"                    "
"                    "
"─ openAI on 4o ─────"
" repeat thisrepeat  "
" thisrepeat this    "
